lib/server/src/server.rs

```diff
@@ -474,65 +478,93 @@ impl TrowServer {
         Ok(())
     }
 
-    async fn get_auth_token(&self, cl: &reqwest::Client, image: &Image, auth: &Auth) -> Option<String> {
+    async fn get_auth_token(&self, cl: &reqwest::Client, image: &Image, auth: &Option<Auth>) -> Result<String, Error> {
+        let www_authenticate_header = self.get_www_authenticate_header(cl, image).await?;
+        debug!("www_authenticate_header: {}",www_authenticate_header);
+        self.get_auth_token_from_header(cl, auth, www_authenticate_header).await
+    }
 
-        //Fucking annoyingly, anonymous requests have to use GET whilst user based requests have to be POST
+    async fn get_auth_token_from_header(&self, cl: &reqwest::Client, auth: &Option<Auth>, www_authenticate_header: String) -> Result<String, Error> {
+        let mut bearer_param_map = TrowServer::get_bearer_param_map(www_authenticate_header);
 
-        let resp = if !&auth.pass.is_some() {
-            info!("Making anonymous auth request to Docker Hub");
-            let req = format!("{}?service={}&scope=repository:{}:pull", &auth.endpoint, auth.resource, image.repo).to_string();
+        let realm = bearer_param_map.get("realm")
+            .cloned()
+            .ok_or(format_err!("Realm should exists in Bearer header"))?;
 
-            cl.get(&req).send().await
-        } else {            
-            let mut auth_req = format!("&grant_type=password&client_id=trow_proxy&service={}&scope=repository:{}:pull", auth.resource, image.repo);
+        bearer_param_map.remove("realm");
 
-            if let Some(user) = &auth.user {
-                info!("Making request to Docker Hub as {}", &user);
-                auth_req = format!("{}&username={}", auth_req, user);
-            }
-            if let Some(pass) = &auth.pass {
-                auth_req = format!("{}&password={}", auth_req, pass);
-            }
+        let reqBuilder = cl.get(realm.as_str())
+            .query(
+                &bearer_param_map
+            );
 
-            cl.post(&auth.endpoint)
-                .header("Content-Type", "application/x-www-form-urlencoded")
-                .body(auth_req).send().await
+        let reqBuilder = match auth {
+            Some(a) => {
+                if a.user.is_some() {
```

Comment:
I think this can be refactored to get rid of the unwraps. Something like `if let Some(a_user) = &a.user {`

Then use `a_user` instead of `a.user`